awscli/customizations/cloudfront.py

```diff
@@ -50,14 +63,61 @@ def __init__(self):
 
     def add_to_params(self, parameters, value):
         if value is not None:
-            caller_reference = 'cli-%s-%s' % (
-                int(time.time()), random.randint(1, 1000000))
             parameters['InvalidationBatch'] = {
-                "CallerReference": caller_reference,
+                "CallerReference": caller_reference(),
                 "Paths": {"Quantity": len(value), "Items": value},
                 }
 
 
+def _add_origin_domain_name(argument_table, **kwargs):
+    argument_table['distribution-config'].required = False
+    argument_table['origin-domain-name'] = OriginDomainNameArgument()
+
+
+class OriginDomainNameArgument(CustomArgument):
+
+    def __init__(self):
+        doc = (
+            'The domain name for your origin. Note: --origin-domain-name '
+            'and --distribution-config are mututally exclusive.'
+        )
+        super(OriginDomainNameArgument, self).__init__(
+            'origin-domain-name', help_text=doc)
+
+    def add_to_params(self, parameters, value):
+        if value is None:
+            return
+        origin_id = caller_reference(prefix='origin_id')
+        item = {"Id": origin_id, "DomainName": value, "OriginPath": ""}
```

Comment:
Deriving from domain name looks better, but there is a risk of duplicated ID. In the console, duplication is less a problem because it will provide a visual error message and the customer just need to type something else in the input box. But in CLI, there is no easy way for customer to get around this. So I would suggest to just use a basically random string here. Feel free to let me know if there is better solution.